Test/lambdas/StateMonad.dfy

```diff
@@ -0,0 +1,231 @@
+// RUN: %dafny /compile:0 /dprint:"%t.dprint" "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+abstract module Monad {
+  type M<A>
+
+  function method Return<A>(x: A): M<A>
+  function method Bind<A,B>(m: M<A>, f:A -> M<B>): M<B>
+
+  // return x >>= f = f x
+  lemma LeftIdentity<A,B>(x : A, f : A -> M<B>)
+    ensures Bind(Return(x),f) == f(x)
+
+  // m >>= return = m
+  lemma RightIdentity<A>(m : M<A>)
+    ensures Bind(m, x => Return(x)) == m // problem: can't prove it for Bind(m, Return)
+
+  // (m >>= f) >>= g = m >>= (x => f(x) >>= g)
+  lemma Associativity<A,B,C>(m : M<A>, f:A -> M<B>, g: B -> M<C>)
+    ensures Bind(Bind(m,f),g) ==
+            Bind(m,x => Bind(f(x),g))
+}
+
+module State refines Monad {
+//   type S // doesn't work because S is used left of an arrow
+  datatype M<A> = State(runState: int -> (A, int))
+
+  function method Return<A>(x: A): M<A>
+  { State(s => (x, s)) }
+
+  function method Bind<A,B>(m: M<A>, f:A -> M<B>): M<B>
+  {
+    match m case State(h) =>
+      State(s =>
+            match h(s) case (a, newState) =>
+              match f(a) case State(g) => g(newState))
+  }
+
+  lemma FunEq<X, Y>(f: X -> Y, g: X -> Y)
+    requires forall x :: f(x) == g(x) 
+    ensures f == g 
+  { assume false; }
+
+  lemma LeftIdentity<A,B>(x : A, f : A -> M<B>)
+    ensures Bind(Return(x), f) == f(x) // return x >>= f  ==  f x
```

Comment:
1 space around == in comment

Reply:
I removed that comment now since it's repeating another comment from above.